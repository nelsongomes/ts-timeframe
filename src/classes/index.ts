export { Timeline } from './timeline';
export { TimelineEvent } from './timeline-event';
