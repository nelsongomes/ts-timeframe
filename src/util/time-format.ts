import { TimelineTimestamp } from '../types/timeline-types';
import {
  getDeltaMicroseconds,
  getDeltaMilliseconds,
  getDeltaNanoseconds,
  getDeltaSeconds,
  round
} from './time';

// #region delta functions readable
/**
 * Time expressed in seconds
 * @param start time
 * @param end time
 * @param decimals optional number of decimals (default 3)
 * Returns formatted string
 */
export function getReadableDeltaSeconds(
  start: TimelineTimestamp,
  end: TimelineTimestamp,
  decimals: number = 3
): string {
  return `${getDeltaSeconds(start, end).toFixed(decimals)}`;
}

/**
 * Time expressed in milliseconds
 * @param start time
 * @param end time
 * @param decimals optional number of decimals (default 3)
 * Returns formatted string
 */
export function getReadableDeltaMilliseconds(
  start: TimelineTimestamp,
  end: TimelineTimestamp,
  decimals: number = 3
): string {
  return `${getDeltaMilliseconds(start, end).toFixed(decimals)}`;
}

/**
 * Time expressed in microseconds
 * @param start time
 * @param end time
 * @param decimals optional number of decimals (default 3)
 * Returns formatted string
 */
export function getReadableDeltaMicroseconds(
  start: TimelineTimestamp,
  end: TimelineTimestamp,
  decimals: number = 3
): string {
  return `${round(getDeltaMicroseconds(start, end), 3).toFixed(decimals)}`;
}

/**
 * Time expressed in microseconds
 * @param start time
 * @param end time
 * @param decimals optional number of decimals (default 3)
 * Returns formatted string
 */
export function getReadableDeltaNanoseconds(
  start: TimelineTimestamp,
  end: TimelineTimestamp,
  decimals: number = 3
): string {
  return `${round(getDeltaNanoseconds(start, end), 3).toFixed(decimals)}`;
}
// #endregion
