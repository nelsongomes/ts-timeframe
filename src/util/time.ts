import { ITimelineUnit, TimelineTimestamp } from "../types/timeline-types";

// #region delta functions

/**
 * Method returns time difference expressed as a string in the unit selected, with n decimal places
 * @param start time
 * @param end time
 * @param unit unit to express result
 * @param decimals
 */
export function getDeltaUnitsReadable(
  start: TimelineTimestamp,
  end: TimelineTimestamp,
  unit: ITimelineUnit,
  decimals: number
): string {
  return getDeltaUnits(start, end, unit).toFixed(decimals) + ` ${unit}`;
}

export function getDeltaUnits(
  start: TimelineTimestamp,
  end: TimelineTimestamp,
  unit: ITimelineUnit
): number {
  switch (unit) {
    case ITimelineUnit.Seconds:
      return getDeltaSeconds(start, end);
    case ITimelineUnit.Milliseconds:
      return getDeltaMilliseconds(start, end);
    case ITimelineUnit.Microseconds:
      return getDeltaMicroseconds(start, end);
    case ITimelineUnit.Nanoseconds:
      return getDeltaNanoseconds(start, end);
  }
}

/**
 * Method returns number of seconds between 2 times expressed in seconds (with floating point precision)
 * @param start time
 * @param end time
 */
export function getDeltaSeconds(
  start: TimelineTimestamp,
  end: TimelineTimestamp
): number {
  const diff = delta(start, end);
  return diff[0] + diff[1] / 1e9;
}

/**
 * Method returns number of milliseconds between 2 times expressed in milliseconds (with floating point precision)
 * @param start time
 * @param end time
 */
export function getDeltaMilliseconds(
  start: TimelineTimestamp,
  end: TimelineTimestamp
): number {
  const diff = delta(start, end);
  return diff[0] * 1e3 + diff[1] / 1e6;
}

/**
 * Method returns number of microseconds between 2 times expressed in microseconds (with floating point precision)
 * @param start time
 * @param end time
 */
export function getDeltaMicroseconds(
  start: TimelineTimestamp,
  end: TimelineTimestamp
): number {
  const diff = delta(start, end);
  return diff[0] * 1e6 + diff[1] / 1e3;
}

/**
 * Method returns number of nanoseconds between 2 times expressed in nanos (with floating point precision)
 * @param start time
 * @param end time
 */
export function getDeltaNanoseconds(
  start: TimelineTimestamp,
  end: TimelineTimestamp
): number {
  const diff = delta(start, end);
  return diff[0] * 1e9 + diff[1];
}

/**
 * Calculates the difference between 2 time structures
 * @param start
 * @param end
 */
export function delta(
  start: TimelineTimestamp,
  end: TimelineTimestamp
): TimelineTimestamp {
  const result: TimelineTimestamp = [end[0] - start[0], end[1] - start[1]];

  if (result[1] < 0) {
    result[0]--;
    result[1] = result[1] + 1e9;
  }

  return result;
}
// #endregion

// #region help functions
/**
 * Returns current time
 */
export function now(): TimelineTimestamp {
  return process.hrtime();
}

/**
 * Delays execution a few milliseconds
 * @param milliseconds
 */
export async function delay(milliseconds: number): Promise<void> {
  return new Promise<void>(resolve => {
    setTimeout(resolve, milliseconds);
  });
}
// #endregion
