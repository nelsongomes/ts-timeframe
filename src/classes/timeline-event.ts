import { TimelineTimestamp } from '../types/timeline-types';
import { now } from '../util/time';

export class TimelineEvent {
  private startTime: TimelineTimestamp;
  private endTime?: TimelineTimestamp;
  private labels: string[];

  constructor(labels: string[]) {
    this.startTime = now();
    this.labels = labels;
  }

  public getStart(): TimelineTimestamp {
    return this.startTime;
  }

  public getEnd(): TimelineTimestamp | undefined {
    return this.endTime;
  }

  public end(): void {
    this.endTime = now();
  }
}
