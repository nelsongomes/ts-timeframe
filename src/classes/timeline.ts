import {
  ITimelineDuration,
  ITimelineSettings,
  ITimelineUnit,
  TimelineTimestamp
} from '../types/timeline-types';
import {
  delta,
  getDeltaMicroseconds,
  getDeltaMilliseconds,
  getDeltaNanoseconds,
  getDeltaSeconds,
  getDeltaUnitsReadable,
  now
} from '../util/time';
import { TimelineEvent } from './timeline-event';

// Timeline configuration
const settings: ITimelineSettings = {
  precision: 3,
  unit: ITimelineUnit.Milliseconds
};

export class Timeline {
  /**
   * Method to set default settings, can only be called once
   * @param newSettings
   */
  public static init(newSettings: Partial<ITimelineSettings>) {
    // init lib once
    Object.assign(settings, newSettings);
    // lock from further inits
    Object.freeze(settings);
  }

  /**
   * Method to get current settings
   */
  public static getDefaults(): ITimelineSettings {
    return settings;
  }

  private timeLineEvents: TimelineEvent[];
  private unit: ITimelineUnit;
  private precision: number;

  /**
   * Class constructor
   * @param defaultUnit
   * @param defaultPrecision
   */
  constructor(
    defaultUnit: ITimelineUnit = settings.unit,
    defaultPrecision: number = settings.precision
  ) {
    // copy settings
    this.unit = defaultUnit;
    this.precision = defaultPrecision;

    // mark the start with a new event
    this.timeLineEvents = [];
    this.innerStart();
  }

  public end() {
    if (!this.timeLineEvents[0].getEnd()) {
      this.timeLineEvents[0].end();
    }

    // check if there are any slow calls (this is done here to minimize performance impacts)
    this.executeSlownessCallbacks();
  }

  public duration(): ITimelineDuration {
    const start = this.timeLineEvents[0].getStart();
    const end = this.timeLineEvents[0].getEnd();

    if (!end) {
      throw new Error('end() function was not called...');
    }

    let duration: number;

    switch (this.unit) {
      case ITimelineUnit.Microseconds:
        duration = getDeltaMicroseconds(start, end);
        break;
      case ITimelineUnit.Milliseconds:
        duration = getDeltaMilliseconds(start, end);
        break;
      case ITimelineUnit.Nanoseconds:
        duration = getDeltaNanoseconds(start, end);
        break;
      case ITimelineUnit.Seconds:
        duration = getDeltaSeconds(start, end);
        break;
      default:
        throw new Error('This should not happen');
    }

    return { unit: this.unit, duration };
  }

  /**
   * Returns the events count.
   */
  public count(): number {
    // first event is timeline creation event itself
    return this.timeLineEvents.length - 1;
  }

  /**
   * Sum duration of all events
   */
  public sumEventsDuration(): TimelineTimestamp {
    const sum: TimelineTimestamp = [0, 0];

    let i = 0;
    for (const iEvent of this.timeLineEvents) {
      if (i > 0) {
        const duration = iEvent.getDurationRaw();

        if (duration) {
          sum[0] += duration[0];
          sum[1] += duration[1];

          while (sum[1] > 1e9) {
            sum[0]++;
            sum[1] -= 1e9;
          }
        }
      }
      i++;
    }

    return sum;
  }

  public startEvent(labels: string[] = []): TimelineEvent {
    if (this.timeLineEvents[0].getEnd()) {
      throw new Error('Cannot start more events after timeline.end()');
    }

    const event: TimelineEvent = new TimelineEvent(labels);

    // add event to timeline
    this.timeLineEvents.push(event);

    return event;
  }

  /**
   * Returns a string with analytic information for this timeline
   */
  public generateAnalyticInfo(): string {
    if (!this.timeLineEvents[0].getEnd()) {
      throw new Error('end() function was not called...');
    }

    let output = `***** Analytic Information for this Timeline *****\n`;

    output += `Total events: ${this.count()}\n`;
    output += `Grand duration: ${getDeltaUnitsReadable(
      this.timeLineEvents[0].getStart(),
      this.timeLineEvents[0].getEnd() as TimelineTimestamp,
      this.unit,
      this.precision
    )}\n`;
    output += `Events duration: ${getDeltaUnitsReadable(
      [0, 0],
      this.sumEventsDuration(),
      this.unit,
      this.precision
    )}\n`;

    if (this.count()) {
      output += this.generateDetailedAnalyticInfo();
    }

    return output;
  }

  /**
   * Returns a string with detailed analytic information for this timeline
   */
  public generateDetailedAnalyticInfo(): string {
    const offset: TimelineTimestamp = this.timeLineEvents[0].getStart();
    let output = `\n***** Event Detail for this Timeline *****\n`;

    let i = 0;
    for (const event of this.timeLineEvents) {
      if (i > 0) {
        output += `#${i}: [started: ${getDeltaUnitsReadable(
          offset,
          this.timeLineEvents[i].getStart(),
          this.unit,
          this.precision
        )}, duration: ${getDeltaUnitsReadable(
          this.timeLineEvents[i].getStart(),
          this.timeLineEvents[i].getEnd() as TimelineTimestamp,
          this.unit,
          this.precision
        )}]\n`;
      }
      i++;
    }

    return output;
  }

  /**
   * Creates the first event, marking the timeline start
   */
  private innerStart() {
    this.timeLineEvents.push(new TimelineEvent([]));
  }

  private executeSlownessCallbacks(): void {
    // if we have slow event rules and events
    if (settings.slowEvents && settings.slowEvents.length && this.count()) {
      for (let i = 1; i < this.timeLineEvents.length; i++) {
        const currentEvent = this.timeLineEvents[i];

        // check all rules
        for (const slowEvent of settings.slowEvents) {
          const eventDuration = currentEvent.getDuration(this.unit);

          if (
            slowEvent.rule.duration > 0 &&
            eventDuration &&
            eventDuration >= slowEvent.rule.duration &&
            slowEvent.rule.matchAnylabel.some(member => {
              return currentEvent.getLabels().includes(member);
            })
          ) {
            slowEvent.callback(null, {
              ...slowEvent.rule,
              unit: settings.unit
            });
          }
        }
      }
    }
  }
}
