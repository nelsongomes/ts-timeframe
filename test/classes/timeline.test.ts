import { Timeline } from '../../src/classes/timeline';
import {
  ITimelineSettings,
  ITimelineUnit
} from '../../src/types/timeline-types';
import * as time from '../../src/util/time';

const initParameters: ITimelineSettings = {
  precision: 5,
  slowEvents: [
    {
      callback: jest.fn(() => {
        /* some action */
      }),
      rule: {
        duration: 100,
        matchAnylabel: ['database'],
        message: 'Database too slow'
      }
    },
    {
      callback: jest.fn(() => {
        /* some action */
      }),
      rule: {
        duration: 3000,
        matchAnylabel: ['api'],
        message: 'Api calls too slow'
      }
    }
  ],
  unit: ITimelineUnit.Microseconds
};

describe('Timeline', () => {
  describe('should use constructor', () => {
    describe('without Timeline init', () => {
      test('properly ended, with default constructor unit', async () => {
        const timeline = new Timeline();

        expect(timeline instanceof Timeline).toBe(true);
        expect(timeline.count()).toBe(0);

        // terminate timeline
        timeline.end();

        expect(timeline.duration().duration).toBeGreaterThan(0);
        expect(timeline.duration().unit).toBe(ITimelineUnit.Milliseconds);
      });

      test('properly ended, with unit seconds', async () => {
        const timeline = new Timeline(ITimelineUnit.Seconds, 0);

        expect(timeline instanceof Timeline).toBe(true);
        expect(timeline.count()).toBe(0);

        // terminate timeline
        timeline.end();

        expect(timeline.duration().duration).toBeGreaterThan(0);
        expect(timeline.duration().unit).toBe(ITimelineUnit.Seconds);
      });

      test('properly ended, with unit milliseconds', async () => {
        const timeline = new Timeline(ITimelineUnit.Milliseconds);

        await time.delay(50);

        expect(timeline instanceof Timeline).toBe(true);
        expect(timeline.count()).toBe(0);

        // terminate timeline
        timeline.end();

        expect(timeline.duration().duration).toBeGreaterThan(50);
        expect(timeline.duration().unit).toBe(ITimelineUnit.Milliseconds);
      });

      test('properly ended, with unit microseconds', async () => {
        const timeline = new Timeline(ITimelineUnit.Microseconds);

        expect(timeline instanceof Timeline).toBe(true);
        expect(timeline.count()).toBe(0);

        // terminate timeline
        timeline.end();

        expect(timeline.duration().duration).toBeGreaterThan(0);
        expect(timeline.duration().unit).toBe(ITimelineUnit.Microseconds);
      });

      test('properly ended, with unit nanoseconds', async () => {
        const timeline = new Timeline(ITimelineUnit.Nanoseconds);

        expect(timeline instanceof Timeline).toBe(true);
        expect(timeline.count()).toBe(0);

        // terminate timeline
        timeline.end();

        expect(timeline.duration().duration).toBeGreaterThan(0);
        expect(timeline.duration().unit).toBe(ITimelineUnit.Nanoseconds);
      });

      test('properly ended, with invalid unit', async () => {
        const timeline = new Timeline('xxx' as ITimelineUnit);

        expect(timeline instanceof Timeline).toBe(true);
        expect(timeline.count()).toBe(0);

        // terminate timeline
        timeline.end();

        try {
          timeline.duration();
          fail('Should have thrown an error');
        } catch (e) {
          expect((e as Error).message).toBe('This should not happen');
        }
      });

      test('unproperly ended', async () => {
        const timeline = new Timeline();

        try {
          timeline.duration();
          fail();
        } catch (error) {
          expect((error as Error).message).toBe(
            'end() function was not called...'
          );
        }
      });

      test('should throw an error if new event is added after timeline.end()', async () => {
        const timeline = new Timeline();

        try {
          timeline.end();
          timeline.startEvent();
          fail();
        } catch (error) {
          expect((error as Error).message).toBe(
            'Cannot start more events after timeline.end()'
          );
        }
      });
    });
  });

  describe('initialization', () => {
    test('should work first init', async () => {
      Timeline.init(initParameters);

      const timeline = new Timeline();
      timeline.end();

      // check init parameters set
      expect(Timeline.getDefaults()).toMatchObject(initParameters);
    });

    test('should fail second init', async () => {
      try {
        Timeline.init({
          precision: 5,
          unit: ITimelineUnit.Microseconds
        });
        fail();
      } catch (e) {
        // check init parameters unchanged
        expect(Timeline.getDefaults()).toMatchObject(initParameters);
      }
    });
  });

  describe('slow events must trigger a callback on end()', () => {
    beforeEach(() => {
      jest.clearAllMocks();
    });

    test('no events', async () => {
      const spy = jest
        .spyOn(time, 'now')
        .mockReturnValueOnce([0, 0]) // first call
        .mockReturnValueOnce([0, 99999]); // second call

      const timeline = new Timeline();
      timeline.end();

      if (initParameters.slowEvents && initParameters.slowEvents[0]) {
        expect(initParameters.slowEvents[0].callback).toBeCalledTimes(0);
      } else {
        fail();
      }
      expect(spy).toBeCalledTimes(2);
    });

    test('slow event, not matching label', async () => {
      const spy = jest
        .spyOn(time, 'now')
        .mockReturnValueOnce([0, 0])
        .mockReturnValueOnce([0, 99999])
        .mockReturnValueOnce([0, 199999])
        .mockReturnValueOnce([0, 999999]);

      const timeline = new Timeline();
      const event = timeline.startEvent(['wronglabel']);
      event.end();

      timeline.end();

      if (initParameters.slowEvents && initParameters.slowEvents[0]) {
        expect(initParameters.slowEvents[0].callback).toBeCalledTimes(0);
      } else {
        fail();
      }
      expect(spy).toBeCalledTimes(4);
    });

    test('slow event, matching label', async () => {
      const spy = jest
        .spyOn(time, 'now')
        .mockReturnValueOnce([0, 0])
        .mockReturnValueOnce([0, 99999])
        .mockReturnValueOnce([0, 199999])
        .mockReturnValueOnce([0, 999999]);

      const timeline = new Timeline();
      const event = timeline.startEvent(['database', 'mysql']);
      event.end();

      timeline.end();

      if (initParameters.slowEvents && initParameters.slowEvents[0]) {
        expect(initParameters.slowEvents[0].callback).toBeCalledTimes(1);
      } else {
        fail();
      }
      expect(spy).toBeCalledTimes(4);
    });
  });

  describe('should generate Timeline analytics', () => {
    beforeEach(() => {
      jest.clearAllMocks();
      jest.resetAllMocks();
    });

    test('with no events', async () => {
      const spy = jest
        .spyOn(time, 'now')
        .mockReturnValueOnce([0, 0]) // first call
        .mockReturnValueOnce([0, 99999]); // second call

      const timeline = new Timeline();
      timeline.end();

      expect(spy).toBeCalledTimes(2);
      expect(timeline.generateAnalyticInfo()).toMatchSnapshot();
    });

    test('without calling timeline.end()', async () => {
      const spy = jest.spyOn(time, 'now').mockReturnValueOnce([0, 0]); // first call

      const timeline = new Timeline();

      expect(spy).toBeCalledTimes(1);
      try {
        expect(timeline.generateAnalyticInfo()).toMatchSnapshot();
        fail();
      } catch (error) {
        expect((error as Error).message).toBe(
          'end() function was not called...'
        );
      }
    });

    test('one event', async () => {
      const spy = jest
        .spyOn(time, 'now')
        .mockReturnValueOnce([0, 0]) // timeline constructor

        .mockReturnValueOnce([0, 99999]) // startEvent()
        .mockReturnValueOnce([0, 199999]) // endEvent

        .mockReturnValueOnce([0, 299999]); // timeline end

      const timeline = new Timeline(ITimelineUnit.Microseconds);
      const event = timeline.startEvent(['database', 'mysql']);
      event.end();

      timeline.end();

      expect(spy).toBeCalledTimes(4);
      expect(timeline.generateAnalyticInfo()).toMatchSnapshot();
    });

    test('multiple events', async () => {
      const spy = jest
        .spyOn(time, 'now')
        .mockReturnValueOnce([0, 0]) // timeline constructor

        .mockReturnValueOnce([0, 0]) // startEvent()
        .mockReturnValueOnce([0, 999999999]) // endEvent

        .mockReturnValueOnce([0, 99999]) // startEvent()
        .mockReturnValueOnce([0, 199999]) // endEvent

        .mockReturnValueOnce([1, 299999]); // timeline end

      const timeline = new Timeline(ITimelineUnit.Microseconds);
      const event = timeline.startEvent();
      event.end();

      const event2 = timeline.startEvent();
      event2.end();

      timeline.end();

      expect(spy).toBeCalledTimes(6);
      expect(timeline.generateAnalyticInfo()).toMatchSnapshot();
    });
  });
});
