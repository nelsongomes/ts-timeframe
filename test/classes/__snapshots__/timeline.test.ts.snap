// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Timeline should generate Timeline analytics multiple events 1`] = `
"***** Analytic Information for this Timeline *****
Total events: 2
Grand duration: 1000299.99900 µs(s)
Events duration: 1000099.99900 µs(s)

***** Event Detail for this Timeline *****
#1: [started: 0.00000 µs(s), duration: 999999.99900 µs(s)]
#2: [started: 99.99900 µs(s), duration: 100.00000 µs(s)]
"
`;

exports[`Timeline should generate Timeline analytics one event 1`] = `
"***** Analytic Information for this Timeline *****
Total events: 1
Grand duration: 299.99900 µs(s)
Events duration: 100.00000 µs(s)

***** Event Detail for this Timeline *****
#1: [started: 99.99900 µs(s), duration: 100.00000 µs(s)]
"
`;

exports[`Timeline should generate Timeline analytics with no events 1`] = `
"***** Analytic Information for this Timeline *****
Total events: 0
Grand duration: 99.99900 µs(s)
Events duration: 0.00000 µs(s)
"
`;
